spec/units/lexer_spec.rb

```diff
@@ -0,0 +1,152 @@
+require 'spec_helper'
+require 'gitsh/lexer'
+
+RSpec::Matchers.define(:produce_tokens) do |expected|
+  match do |actual|
+    @expected = expected.join("\n")
+    @actual = Gitsh::Lexer.new.lex(actual).map(&:to_s).join("\n")
+    values_match? @expected, @actual
+  end
+
+  diffable
+end
+
+describe Gitsh::Lexer do
+  describe '#lex' do
+    it 'recognises space separated words' do
+      expect('foo bar').
+        to produce_tokens ['WORD(foo)', 'SPACE', 'WORD(bar)', 'EOS']
+      expect('!./bin/setup --help').
+        to produce_tokens ['WORD(!./bin/setup)', 'SPACE', 'WORD(--help)', 'EOS']
+    end
+
+    it 'recognises the semicolon operator' do
+      expect('foo;').
+        to produce_tokens ['WORD(foo)', 'SEMICOLON', 'EOS']
+      expect('foo; bar').
+        to produce_tokens ['WORD(foo)', 'SEMICOLON', 'WORD(bar)', 'EOS']
+      expect('foo  ; bar').
+        to produce_tokens ['WORD(foo)', 'SEMICOLON', 'WORD(bar)', 'EOS']
+    end
+
+    it 'recognises the && operator' do
+      expect('foo && bar').
+        to produce_tokens ['WORD(foo)', 'AND', 'WORD(bar)', 'EOS']
+    end
+
+    it 'recognises the || operator' do
+      expect('foo || bar').
+        to produce_tokens ['WORD(foo)', 'OR', 'WORD(bar)', 'EOS']
+    end
+
+    [' ', "\t", "\r", "\n", "\f", '\'', '"', '\\', '$', '#', ';', '&', '|'].each do |char|
+      it "recognises unquoted words containing an escaped #{char.inspect}" do
+        expect("foo\\#{char}bar").
+          to produce_tokens ['WORD(foo)', "WORD(#{char})", 'WORD(bar)', 'EOS']
+      end
+    end
+
+    it 'recognises variables' do
+      expect('$foo').to produce_tokens ['VAR(foo)', 'EOS']
+      expect('${foo}').to produce_tokens ['VAR(foo)', 'EOS']
+      expect('pre$foo').to produce_tokens ['WORD(pre)', 'VAR(foo)', 'EOS']
+      expect('pre${foo}post').
+        to produce_tokens ['WORD(pre)', 'VAR(foo)', 'WORD(post)', 'EOS']
+      expect('pre$foo/post').
+        to produce_tokens ['WORD(pre)', 'VAR(foo)', 'WORD(/post)', 'EOS']
+      expect('$f_o-o.bar').to produce_tokens ['VAR(f_o-o.bar)', 'EOS']
+      expect('$_bar').to produce_tokens ['VAR(_bar)', 'EOS']
+    end
+
+    it 'recognises single-quoted strings as single words' do
+      expect('\'foo bar\'').
+        to produce_tokens ['WORD(foo bar)', 'EOS']
+    end
+
+    ['\\', '\''].each do |char|
+      it "recognises single-quoted strings containing an escaped #{char.inspect}" do
+        expect("'foo\\#{char}bar'").
+          to produce_tokens ['WORD(foo)', "WORD(#{char})", 'WORD(bar)', 'EOS']
```

Comment:
The job of the lexer is to convert syntax (`\$`) into tokens that are useful in the language (`WORD($)`). The way lexers typically do this is by testing the beginning of the input against a bunch of regular expressions, and going with whichever produces the longest match.

It's be very hard (maybe even impossible?) to write a single regular expression that could handle all of the possible variations of escaping (there's `\$x` to `WORD($x)`, but also more complex things like `\\$x` to `WORD(\) VAR(x)`). Even if it were possible, I think the current set of rules is clearer than whatever we'd need to produce a single `WORD` token for these cases.

There are also other reasons for the parser to combine multiple adjacent `WORD` tokens (see my answer to your comment below about parsing commands broken into multiple words), so it doesn't matter much if the lexer produces `WORD(x) WORD(y)` or `WORD(xy)`.